# Лабораторная работа №6

In [13]:
%load_ext sql
%sql postgresql://AAK:@127.0.0.1/students
%config SqlMagic.displaylimit = None

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


displaylimit: Value None will be treated as 0 (no limit)

1.  ## Задание 1
    > Проведите следующий эксперимент:
    > Создайте простую таблицу, содержащую три поля – суррогатный ключ и два атрибута, содержащие строки.

In [14]:
%%sql
CREATE TABLE Test(
	id SERIAL PRIMARY KEY,
	CODE_1 VARCHAR(64),
	CODE_2 VARCHAR(64)
);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

++
||
++
++

> С помощью следующего скрипта заполните таблицу данными.

In [15]:
%%sql
DO
$$
BEGIN
FOR i IN 1..1000000 LOOP
    INSERT INTO Test(code_1,code_2) VALUES(md5(random()::text), md5(random()::text));
END LOOP;
END
$$ language plpgsql;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

++
||
++
++

> С помощью функций EXPLAIN ANALYZE в PostgreSQL возможно производить анализ запросов и вычислять затрачиваемое время на его выполнение. Например,

In [16]:
%%sql
EXPLAIN ANALYZE SELECT * FROM student;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

3 rows affected.

+------------------------------------------------------------------------------------------------------+
|                                              QUERY PLAN                                              |
+------------------------------------------------------------------------------------------------------+
| Seq Scan on student  (cost=0.00..4.63 rows=163 width=90) (actual time=0.017..0.057 rows=189 loops=1) |
|                                       Planning Time: 0.057 ms                                        |
|                                       Execution Time: 0.069 ms                                       |
+------------------------------------------------------------------------------------------------------+

> Добавьте в таблицу Test одно значение, измерив время данной операции.
> Далее измерьте время выполнения запроса,
> выводящего содержимого таблицы в отсортированном виде по столбцу CODE_1.
>
> Добавьте индекс на столбец CODE_1. Повторите предыдущие две операции.
> Сравните полученное время. Во сколько раз оно изменилось? Результаты вычисления занесите в таблицу.

Добавим одно значение в таблицу:

In [17]:
%%sql
EXPLAIN ANALYZE INSERT INTO Test(code_1,code_2) VALUES(1, 1);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

4 rows affected.

+--------------------------------------------------------------------------------------------+
|                                         QUERY PLAN                                         |
+--------------------------------------------------------------------------------------------+
| Insert on test  (cost=0.00..0.01 rows=0 width=0) (actual time=0.027..0.027 rows=0 loops=1) |
|   ->  Result  (cost=0.00..0.01 rows=1 width=296) (actual time=0.014..0.014 rows=1 loops=1) |
|                                  Planning Time: 0.022 ms                                   |
|                                  Execution Time: 0.035 ms                                  |
+--------------------------------------------------------------------------------------------+

Измерим время выполнения запроса:

In [18]:
%%sql
EXPLAIN ANALYZE SELECT * FROM Test ORDER BY CODE_1;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

11 rows affected.

+------------------------------------------------------------------------------------------------------------------------------------+
|                                                             QUERY PLAN                                                             |
+------------------------------------------------------------------------------------------------------------------------------------+
|        Gather Merge  (cost=43129.77..73139.47 rows=257208 width=296) (actual time=1297.281..2451.633 rows=1000001 loops=1)         |
|                                                          Workers Planned: 2                                                        |
|                                                         Workers Launched: 2                                                        |
|            ->  Sort  (cost=42129.75..42451.26 rows=128604 width=296) (actual time=1266.143..1696.937 rows=333334 loops=3)          |
|                                                              Sort Key: code_1                                                      |
|                                                 Sort Method: external merge  Disk: 26872kB                                         |
|                                           Worker 0:  Sort Method: external merge  Disk: 25704kB                                    |
|                                           Worker 1:  Sort Method: external merge  Disk: 25704kB                                    |
|         ->  Parallel Seq Scan on test  (cost=0.00..13632.04 rows=128604 width=296) (actual time=0.019..30.895 rows=333334 loops=3) |
|                                                      Planning Time: 0.058 ms                                                       |
|                                                    Execution Time: 2477.770 ms                                                     |
+------------------------------------------------------------------------------------------------------------------------------------+

Добавим индекс:

In [19]:
%%sql
CREATE INDEX code_1_idx ON Test (code_1);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

++
||
++
++

Повторим операции:

In [20]:
%%sql
EXPLAIN ANALYZE INSERT INTO Test(code_1,code_2) VALUES(1, 1);

Running query in 'postgresql://AAK:***@127.0.0.1/students'

4 rows affected.

+--------------------------------------------------------------------------------------------+
|                                         QUERY PLAN                                         |
+--------------------------------------------------------------------------------------------+
| Insert on test  (cost=0.00..0.01 rows=0 width=0) (actual time=1.191..1.191 rows=0 loops=1) |
|   ->  Result  (cost=0.00..0.01 rows=1 width=296) (actual time=0.004..0.004 rows=1 loops=1) |
|                                  Planning Time: 0.023 ms                                   |
|                                  Execution Time: 1.200 ms                                  |
+--------------------------------------------------------------------------------------------+

*Execution time* увеличился

In [21]:
%%sql
EXPLAIN ANALYZE SELECT * FROM Test ORDER BY CODE_1;

Running query in 'postgresql://AAK:***@127.0.0.1/students'

3 rows affected.

+-------------------------------------------------------------------------------------------------------------------------------------+
|                                                              QUERY PLAN                                                             |
+-------------------------------------------------------------------------------------------------------------------------------------+
| Index Scan using code_1_idx on test  (cost=0.42..93224.44 rows=1000001 width=296) (actual time=0.027..451.414 rows=1000002 loops=1) |
|                                                       Planning Time: 0.060 ms                                                       |
|                                                      Execution Time: 467.557 ms                                                     |
+-------------------------------------------------------------------------------------------------------------------------------------+

*Execution time* значительно сократился

**Вывод**: добавлять записи дольше, а искать быстрее.

|                    | insert | select |
| ------------------ | ------ | ------ |
| До индекса (мс)    | 0.035  | 2477.7 |
| После индекса (мс) | 1.2    | 467.5  |
| Прирост (раз)      | 0.0291 | 5.23   |